cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/HttpRequestRouter.java

```diff
@@ -206,7 +207,16 @@ private Discoverable getDiscoverable(HttpRequest httpRequest) {
       throw new HandlerException(HttpResponseStatus.SERVICE_UNAVAILABLE,
                                  "No endpoint strategy found for request " + getRequestLine(httpRequest));
     }
+    // Do a non-blocking pick first. If the service has been discovered before, this should return an endpoint
+    // immediately.
```

Comment:
doesn't this make it sticky - it will always use what has been discovered before. If there are new instances to discover, it will not discover them.

Reply:
The `ServiceDiscovered` maintains a live `Iterator` that is used by the `pick` method, so it is not sticky.